File: c/key-gen-face-five.c
#include <stdio.h>
#include <time.h>

int main()
{
    clock_t clock0, clock1, clocki; // clock_t is defined in <time.h> as int
    int c1, c2, c3, c4, c5;         // test faces running from 0 to 12
    int t, k, validKey;             // t=trial key, k=index searched key, validKey=boolean attribute to t (1=valid)
    int i, j;                       // indexes running through sums
    int sums[50000], c;             // array of all possible sums of key[c1-5]

    int key[13] = {0, 1, 5, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0}; // init keys - empirical
    k = 3;
    printf("bootstrap: keys = ");
    for (i = 0; i < k; i++)
        printf("%d ", key[i]);
    printf("\n");

    clock0 = clock();

    printf("searching keys from k =%d\n", k);
    // while (k <= 12)
    while (k <= 10)
    { //------------------choose nb keys to search (<=12)
        t = key[k - 1] + 1;
        clocki = clock();
        do
        {
            key[k] = t;
            validKey = 1;
            c = 0;

            for (c1 = 0; c1 <= k; c1++)
            {
                for (c2 = c1; c2 <= k; c2++)
                {
                    for (c3 = c2; c3 <= k; c3++)
                    {
                        for (c4 = c3; c4 <= k; c4++)
                        {
                            for (c5 = c4; c5 <= k; c5++)
                            {
                                for (c5 = c4; c5 <= k; c5++)
                                {
                                    if (c1 != c5)
                                    {
                                        sums[c] = key[c1] + key[c2] + key[c3] + key[c4] + key[c5];
                                        c++;
                                    }
                                }
                            }
                        }
                    }
                }
            }

            i = 0;
            do
            {
                j = i + 1;
                do
                {
                    if (sums[i] == sums[j])
                    {
                        validKey = 0;
                    }
                    j++;
                } while (validKey == 1 && j < c);
                i++;
            } while (validKey == 1 && i < c - 1);

            if (validKey == 1)
            {
                printf("\tkey[%d] = %d\n", k, t);
                clock1 = clock();
                printf("\t\trun time for key[%d] =\t%.2f s\n", k, (float)(clock1 - clocki) / CLOCKS_PER_SEC);
                printf("\t\trun time to key[%d] =\t%.2f s\n", k, (float)(clock1 - clock0) / CLOCKS_PER_SEC);
            }
            else
            {
                t++;
            }
        } while (validKey == 0);
        k++;
    }
    printf("done\n");
    return 0;
}

/*output
gcc -Wall -g -O3 key-gen-face-five.c -o key-gen-face-five-exec
./key-gen-face-five-exec
bootstrap: keys = 0 1 5 
searching keys from k =3
        key[3] = 22
                run time for key[3] =   0.00 s
                run time to key[3] =    0.00 s
        key[4] = 94
                run time for key[4] =   0.00 s
                run time to key[4] =    0.00 s
        key[5] = 312
                run time for key[5] =   0.00 s
                run time to key[5] =    0.00 s
        key[6] = 992
                run time for key[6] =   0.00 s
                run time to key[6] =    0.01 s
        key[7] = 2422
                run time for key[7] =   0.04 s
                run time to key[7] =    0.04 s
        key[8] = 5624
                run time for key[8] =   0.10 s
                run time to key[8] =    0.14 s
        key[9] = 12522
                run time for key[9] =   0.47 s
                run time to key[9] =    0.62 s
        key[10] = 19998
                run time for key[10] =  0.85 s
                run time to key[10] =   1.46 s
        key[11] = 43258
                run time for key[11] =  5.53 s
                run time to key[11] =   7.00 s
        key[12] = 79415
                run time for key[12] =  18.58 s
                run time to key[12] =   25.58 s
done
*/

File: rust/src/key_gen_face_five.rs
use std::cmp;
use std::time::Instant;

pub fn main() {
    println!("start key-gen-face-five");

    // init
    let mut t: u32; // t=trial key, k=index searched key
    let mut c: usize; // sums counter
    let mut valid: bool; // true if key is valid

    let mut sums = [0; 50000]; // array of all possible sums of key[c[1-5]]
    let mut key = [0, 1, 5, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]; // init keys - empirical

    // let sums_max_idx = sums.len();

    println!("bootstrap -> keys={:?}", key);

    let start = Instant::now();

    for k in 3..11 {
        println!("searching for k={}", k);
        t = key[k - 1] + 1u32;
        let interm = Instant::now();
        valid = false;

        while !valid {
            key[k] = t;
            let c_max = k + 1;
            c = 0;

            for c1 in 0..c_max {
                for c2 in c1..c_max {
                    for c3 in c2..c_max {
                        for c4 in c3..c_max {
                            for c5 in c4..c_max {
                                if c1 != c5 {
                                    // let bounded_c = cmp::min(c, sums_max_idx);
                                    // let bounded_c = cmp::min(c, sums.len() - 1);
                                    let bounded_c = cmp::min(c, sums.len() - 1);
                                    sums[bounded_c] =
                                        key[c1] + key[c2] + key[c3] + key[c4] + key[c5];
                                    c += 1;
                                }
                            }
                        }
                    }
                }
            }

            // for c1 in 0..c_max {
            //     for c2 in c1..c_max {
            //         for c3 in c2..c_max {
            //             for c4 in c3..c_max {
            //                 for c5 in c4..c_max {
            //                     if c1 != c5 {
            //                         unsafe {
            //                             sums[cmp::min(c, sums.len())] = key.get_unchecked(c1)
            //                                 + key.get_unchecked(c2)
            //                                 + key.get_unchecked(c3)
            //                                 + key.get_unchecked(c4)
            //                                 + key.get_unchecked(c5);
            //                             c += 1;
            //                         }
            //                     }
            //                 }
            //             }
            //         }
            //     }
            // }

            // v1 - debug mode
            // let mut i: usize;
            // let mut j: usize;
            // valid = true;
            // i = 0;
            // loop {
            //     j = i + 1;
            //     loop {
            //         if sums[i] == sums[j] {
            //             valid = false;
            //         }
            //         j += 1;

            //         if !(valid && j < c) {
            //             break;
            //         }
            //     }
            //     i += 1;
            //     if !(valid && i < c - 1) {
            //         break;
            //     }
            // }

            // v2 -  release mode with unsafe - fastest
            let mut i: usize;
            let mut j: usize;
            i = 0;
            valid = true;
            loop {
                j = i + 1;
                loop {
                    if unsafe { sums.get_unchecked(i) == sums.get_unchecked(j) } {
                        valid = false;
                    }
                    j += 1;

                    if !(valid && j < c) {
                        break;
                    }
                }
                i += 1;
                if !(valid && i < c - 1) {
                    break;
                }
            }

            // v3 - release mode without unsafe - fastest safe
            // let mut i;
            // i = 0;
            // valid = true;
            // loop {
            //     if sums[i + 1..c].contains(&sums[i]) {
            //         valid = false;
            //     }
            //     i += 1;
            //     if !(valid && i < c - 1) {
            //         break;
            //     }
            // }

            // v3b - release mode without unsafe - fatest safe equivalent more compact
            // valid = (0..c - 1).all(|i| !sums[i + 1..c].contains(&sums[i]));

            // v4 - release mode without unsafe - less efficient ways
            // let mut i: usize;
            // i = 0;
            // valid = true;
            // loop {
            //     let sums_i = sums[i];
            //     for &sums_j in &sums[i + 1..c] {
            //         if sums_i == sums_j {
            //             valid = false;
            //             break;
            //         }
            //     }
            //     i += 1;
            //     if !(valid && i < c - 1) {
            //         break;
            //     }
            // }

            // v5 - release mode without unsafe - less efficient ways
            // valid = true;
            // for (i, s) in sums[..c - 1].iter().enumerate() {
            //     if sums[i + 1..c].contains(s) {
            //         valid = false;
            //         break;
            //     }
            // }

            if valid {
                println!("key[{}]={:?}", k, t);
                // println!("c={:?}", c); // ?!?? comment to allow bounded_c = len()-1 without perf impact
                let end = Instant::now();
                println!("\truntime for key[{}] = {:?}", k, (end - interm));
                println!("\truntime for key[{}] = {:?}", k, (end - start));
            } else {
                t += 1;
            }
        }
    }

    let end = Instant::now();
    println!("runtime = {:?}", (end - start));
    println!("key={:?}", key);
}
